From: Liam Howlett <liam.howlett@oracle.com>
Subject: mm/mmap: fix return on maple tree expand fail in brk()
Date: Tue, 12 Jul 2022 14:51:08 +0000

When do_brk_flags() fails, ensure the necessary locking is undone.

Link: https://lkml.kernel.org/r/20220712145059.4191132-1-Liam.Howlett@oracle.com
Fixes: e9bad648fe05 (mm/mmap: change do_brk_flags() to expand existing VMA and add do_brk_munmap())
Signed-off-by: Liam R. Howlett <Liam.Howlett@oracle.com>
Suggested-by: Hugh Dickins <hughd@google.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/mmap.c |    9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

--- a/mm/mmap.c~mm-mmap-change-do_brk_flags-to-expand-existing-vma-and-add-do_brk_munmap-fix
+++ a/mm/mmap.c
@@ -2965,7 +2965,7 @@ static int do_brk_flags(struct ma_state
 		vma->vm_end = addr + len;
 		vma->vm_flags |= VM_SOFTDIRTY;
 		if (mas_store_gfp(mas, vma, GFP_KERNEL))
-			return -ENOMEM;
+			goto mas_expand_failed;
 
 		if (vma->anon_vma) {
 			anon_vma_interval_tree_post_update_vma(vma);
@@ -3013,6 +3013,13 @@ mas_store_fail:
 vma_alloc_fail:
 	vm_unacct_memory(len >> PAGE_SHIFT);
 	return -ENOMEM;
+
+mas_expand_failed:
+	if (vma->anon_vma) {
+		anon_vma_interval_tree_post_update_vma(vma);
+		anon_vma_unlock_write(vma->anon_vma);
+	}
+	return -ENOMEM;
 }
 
 int vm_brk_flags(unsigned long addr, unsigned long request, unsigned long flags)
_
